Sources/GRPCNIOTransportCore/GRPCStreamStateMachine.swift

```diff
@@ -659,7 +659,8 @@ extension GRPCStreamStateMachine {
     }
 
     for metadataPair in customMetadata {
-      headers.add(name: metadataPair.key, value: metadataPair.value.encoded())
+      // Lowercase the field names for user-provided metadata.
+      headers.add(name: metadataPair.key.lowercased(), value: metadataPair.value.encoded())
```

Comment:
I don't think so; not all transports require lowercase keys (e.g. in-process). What we probably should do is make the _lookup_ by key case insensitive though.

Reply:
Hm I'm not 100% sure about the lookup being case-insensitive if adding isn't. What if someone willingly adds both `somekey` and `SomeKey`, and wants to get the single value defined for each separately? I know this sounds like potentially sketchy decisions from the user, but I feel like the storing and the looking up should probably be symmetrical in this sense.

Reply:
Metadata keys are case-insensitive in gRPC (https://grpc.io/docs/what-is-grpc/core-concepts/#metadata).

Metadata keys aren't required to be one or the other so we shouldn't convert them unnecessarily. In HTTP/2 header field names _are_ required to be lowercase so that's the appropriate time to lowercase them.